lib/spack/spack/build_systems/oneapi.py

```diff
@@ -150,3 +151,7 @@ def link_flags(self):
     @property
     def ld_flags(self):
         return '{0} {1}'.format(self.search_flags, self.link_flags)
+
+    def __add__(self, other):
+        return self.__class__(self.static_libs + other.static_libs,
+                              self.dynamic_libs + other.dynamic_libs)
```

Comment:
This only works if `other` is an `IntelOneApiStaticLibraryList`, I think this will need to use an `isinstance` test.

Also FWIW for X and Y, X is an `IntelOneApiStaticLibraryList` and Y is a `LibraryList`, `X + Y` will generate an `IntelOneApiStaticLibraryList` but `Y + X` will not, which I think would be problematic.

Did you create this PR after running into an error somewhere? If so, can you describe it here (e.g. the package)?

Reply:
Hi @scheibelp, thanks for your comments.

I created this issue after running into problems with `mumps`, `superlu-dist` and `petsc`:

mumps:
```python
lapack_blas = (self.spec['lapack'].libs + self.spec['blas'].libs)
```

superlu-dist:
```python
'-DTPL_BLAS_LIBRARIES=%s' % spec['blas'].libs.joined(";"),
```

petsc:
```python
lapack_blas = spec['lapack'].libs + spec['blas'].libs
options.extend([
    '--with-blas-lapack-lib=%s' % lapack_blas.joined()
])
```

Reply:
Thanks for the examples! I see how that would be problematic.

To address my concerns in https://github.com/spack/spack/pull/27742#discussion_r760725236 I think the following would work (if you don't have the time/interest, let me know and I can make these changes myself and add them to this PR):

* In `llnl/util/filesystem.py` add a `LibraryListComposite` object:
  * It inherits from LibraryList
  * It stores a list of LibraryList objects (e.g. [X, Y])
  * Implement `LibraryListComposite.files` as a property: (return X.files + Y.files)
  * Override search_flags and link_flags (e.g. `LibraryListComposite.search_flags` returns X.search_flags + Y.search_flags)
* Update `LibraryList.add` to return a `LibraryListComposite([self, other])`
* `IntelOneapiStaticLibraryList` can now remove all method definitions except for `search_flags` and `link_flags` (the edits to inherit from LibraryList as added in this PR will be required)

That would restore consistency and also fix the problem you encountered.